### ISSUE TITLE
Does nnUnet work with CrossEntropy alone?

### ISSUE THREAD

**Issue by haochenz23** [MIC-DKFZ/nnUNet#1584]
Im currently experiencing 0 dice score while training on Cross Entropy alone. The loss seems to converge quickly but the inference out all zeros. Any idea what might be the cause of this?

**Comment by FabianIsensee** [comment]
Yes it does, you can easily verify that by runing the Hippocamput dataset from the MSD:
`
nnUNetv2_train 4 3d_fullres 0 -tr nnUNetTrainerCELoss
`

![progress](https://github.com/MIC-DKFZ/nnUNet/assets/11302164/5e1bb751-94ef-47b8-bb3a-dd1d5a80ce00)

(I did not let the training finished but the predictions are not empty with a mean dice of 0.88)

Cross-entropy is not a good loss function for segmentation due to the class imbalance we often have. I would recommend to stick with the default!

**Comment by ykirchhoff** [comment]
Just to add to what Fabian said, it might happen, that the pseudo dice stays at zero for the first few epochs, even with the default ce+dice loss. If you actually get a dice of zero in inference (or continuously during training) there is almost certainly something wrong with your data.

**Comment by haochenz23** [comment]
Here's the mean dice score and loss that I acquired via the default compound loss:
![image](https://github.com/MIC-DKFZ/nnUNet/assets/133702251/98b673af-3119-4145-86b2-e26f8b52756c)
![image](https://github.com/MIC-DKFZ/nnUNet/assets/133702251/62c05240-275a-4d14-8a64-fc6deb12b028)
And here's the mean dice score and loss using Cross Entropy only:
![image](https://github.com/MIC-DKFZ/nnUNet/assets/133702251/df32bcff-8d7f-4c91-b66b-b29ce96a6681)
![image](https://github.com/MIC-DKFZ/nnUNet/assets/133702251/3ec02dbf-b6d2-4c76-b1dd-5a6fb5ad0659)

It confused me that the CE loss seems to be going down but the accuracy is surprisingly zero. I've been using the same data and label for the two runs. The only difference I made to change the loss is in **_build_loss()** function in **nnUNetTrainer.py**:
`        loss = RobustCrossEntropyLoss()` 
and then wrap the **DeepSupervisionWrapper** around it. Am I missing anything else? Thanks!

**Comment by FabianIsensee** [comment]
Can you reproduce this issue with any of the publicly available datasets?

**Comment by haochenz23** [comment]
Yes, this is trained on the ASOCA dataset (https://asoca.grand-challenge.org/), one additional label is added upon the annotations

**Comment by FabianIsensee** [comment]
Can you please share a dataset conversion script for the data? You can make this a PR and get recognition for it

**Comment by haochenz23** [comment]
> Just to add to what Fabian said, it might happen, that the pseudo dice stays at zero for the first few epochs, even with the default ce+dice loss. If you actually get a dice of zero in inference (or continuously during training) there is almost certainly something wrong with your data.

Thanks for the heads up! That's indeed what I observed while training with compound loss. Just out of curiosity is there an explanation for this kind of behavior?

**Comment by haochenz23** [comment]
> Can you please share a dataset conversion script for the data? You can make this a PR and get recognition for it

Sure, I can work on it then it might be easier for you to reproduce the error. The data access is upon request but I think I'm permitted to share it here with a link

**Comment by ykirchhoff** [comment]
> > Just to add to what Fabian said, it might happen, that the pseudo dice stays at zero for the first few epochs, even with the default ce+dice loss. If you actually get a dice of zero in inference (or continuously during training) there is almost certainly something wrong with your data.
> 
> Thanks for the heads up! That's indeed what I observed while training with compound loss. Just out of curiosity is there an explanation for this kind of behavior?

I experienced that mostly with harder tasks, where it probably just takes the network longer to learn features to distinguish foreground from background with a reasonable certainty (keep in mind that for the dice argmax is used). But not sure if that really is the case.

What is the relation between steps and epochs in your plots? It doesn't seem to be the standard nnUNet train/val iterations per epoch.

**Comment by FabianIsensee** [comment]
> > Can you please share a dataset conversion script for the data? You can make this a PR and get recognition for it
> 
> Sure, I can work on it then it might be easier for you to reproduce the error. The data access is upon request but I think I'm permitted to share it here with a link

Well it's not public data if  I have to fill out a form first 🙃 Can you please provide us with a way to reproduce the problem?
Like @ykirchhoff I am wondering a bit about your plots. Have you made any modifications to nnU-Net?

**Comment by FabianIsensee** [comment]
Can you please provide us with a simple way to reproduce your problem? One that relies on publicly available data?

**Comment by haochenz23** [comment]
@FabianIsensee Hi Fab! As for now we are moving on with the compound loss. However Id like to share some of the logging result to see if they make sense. I tried to separate the loss by running the following code under **train_step** function in nnUNetTrainer:

```
with autocast(self.device.type, enabled=True) if self.device.type == 'cuda' else dummy_context():
            output = self.network(data)
            l = self.loss(output, target)

            with torch.no_grad():
                # background cross entropy
                bg_weight = torch.tensor([1.0, 0.0, 0.0]).to(device='cuda')
                bgCE = nn.CrossEntropyLoss(weight=bg_weight)
                bgloss = bgCE(output[0], target[0][:, 0].long())
                print(f"Background CE loss: {bgloss}")

                # lumen cross entropy
                lum_weight = torch.tensor([0.0, 1.0, 0.0]).to(device='cuda')
                lumCE = nn.CrossEntropyLoss(weight=lum_weight)
                lumloss = lumCE(output[0], target[0][:, 0].long())
                print(f"lumen CE loss: {lumloss}")

                # cal cross entropy
                cal_weight = torch.tensor([0.0, 0.0, 1.0]).to(device='cuda')
                calCE = nn.CrossEntropyLoss(weight=cal_weight)
                calloss = calCE(output[0], target[0][:, 0].long())
                print(f"calcification CE loss: {calloss}")

                # pure cross entropy
                ce_temp = nn.CrossEntropyLoss()
                l_CE = ce_temp(output[0], target[0][:, 0].long())
                print(f"Pure CE loss: {l_CE}") 
```
The result graphs looked like this:
![image](https://github.com/MIC-DKFZ/nnUNet/assets/133702251/686bc09d-a093-4912-aa96-ba2a5f8ea467)
![image](https://github.com/MIC-DKFZ/nnUNet/assets/133702251/3e865521-390f-4a3c-950a-8fb590c1a2df)
![image](https://github.com/MIC-DKFZ/nnUNet/assets/133702251/cb46d310-bd71-4101-b34e-92831fbd4d80)

Im not sure if they make sense since its quite noisy and there are only 40 cases in the dataset (or the way I try to separate the loss is incorrect, but pytorch CE is multiclass by nature, applying weight is a trick I came up with). But I also logged the dice soft loss and they have showed strong descent curves like you would have expected on a loss curve despite being noisy.

My theory is that our dataset might be highly imbalanced. The ratio of background pixels to target class pixels is huge (total pixel count for background, label1 and label2 is :[759214773, 1153602, 13065]) therefore the majority of the CE loss came from the background segmentation and soft dice loss is necessary to make up for that.

**Comment by FabianIsensee** [comment]
@haochenz23 do you have any more questions or can we close this issue? I could not have explained it better than @ykirchhoff

**Comment by ykirchhoff** [comment]
I will close this issue now, feel free to open another issue if you encounter any other problems @haochenz23
